File: hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/transaction/EthereumTransaction.java
/*
 * Copyright (C) 2019-2023 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.mirror.common.domain.transaction;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.hedera.mirror.common.converter.AccountIdConverter;
import com.hedera.mirror.common.domain.entity.EntityId;
import jakarta.persistence.Column;
import jakarta.persistence.Convert;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.springframework.data.domain.Persistable;

@AllArgsConstructor(access = AccessLevel.PRIVATE) // For builder
@Builder
@Data
@Entity
@NoArgsConstructor
public class EthereumTransaction implements Persistable<Long> {

    @ToString.Exclude
    private byte[] accessList;

    @ToString.Exclude
    private byte[] callData;

    @Convert(converter = AccountIdConverter.class)
    private EntityId callDataId;

    @ToString.Exclude
    private byte[] chainId;

    @Id
    private long consensusTimestamp;

    @ToString.Exclude
    private byte[] data;

    // persisted in tinybar
    private Long gasLimit;

    // persisted in tinybar
    private byte[] gasPrice;

    @ToString.Exclude
    private byte[] hash;

    // persisted in tinybar
    private byte[] maxFeePerGas;

    // persisted in tinybar
    private Long maxGasAllowance;

    // persisted in tinybar
    private byte[] maxPriorityFeePerGas;

    private Long nonce;

    @Convert(converter = AccountIdConverter.class)
    private EntityId payerAccountId;

    private Integer recoveryId;

    @Column(name = "signature_r")
    @ToString.Exclude
    private byte[] signatureR;

    @Column(name = "signature_s")
    @ToString.Exclude
    private byte[] signatureS;

    @Column(name = "signature_v")
    @ToString.Exclude
    private byte[] signatureV;

    @ToString.Exclude
    private byte[] toAddress;

    private Integer type;

    @ToString.Exclude
    private byte[] value;

    @JsonIgnore
    @Override
    public Long getId() {
        return consensusTimestamp;
    }

    @JsonIgnore
    @Override
    public boolean isNew() {
        return true; // Since we never update and use a natural ID, avoid Hibernate querying before insert
    }
}

File: hedera-mirror-rest/model/ethereumTransaction.js
/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2022 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import _ from 'lodash';

class EthereumTransaction {
  /**
   * Parses ethereum_transaction table columns into object
   */
  constructor(ethereumTransaction) {
    Object.assign(
      this,
      _.mapKeys(ethereumTransaction, (v, k) => _.camelCase(k))
    );
  }

  static tableAlias = 'etht';
  static tableName = 'ethereum_transaction';

  static ACCESS_LIST = 'access_list';
  static CALL_DATA_ID = 'call_data_id';
  static CALL_DATA = 'call_data';
  static CHAIN_ID = 'chain_id';
  static CONSENSUS_TIMESTAMP = 'consensus_timestamp';
  static DATA = 'data';
  static FROM_ADDRESS = 'from_address';
  static GAS_LIMIT = 'gas_limit';
  static GAS_PRICE = 'gas_price';
  static HASH = 'hash';
  static MAX_FEE_PER_GAS = 'max_fee_per_gas';
  static MAX_GAS_ALLOWANCE = 'max_gas_allowance';
  static MAX_PRIORITY_FEE_PER_GAS = 'max_priority_fee_per_gas';
  static NONCE = 'nonce';
  static PAYER_ACCOUNT_ID = 'payer_account_id';
  static RECOVERY_ID = 'recovery_id';
  static SIGNATURE_R = 'signature_r';
  static SIGNATURE_S = 'signature_s';
  static SIGNATURE_V = 'signature_v';
  static TO_ADDRESS = 'to_address';
  static TYPE = 'type';
  static VALUE = 'value';

  /**
   * Gets full column name with table alias prepended.
   *
   * @param {string} columnName
   * @private
   */
  static getFullName(columnName) {
    return `${this.tableAlias}.${columnName}`;
  }
}

export default EthereumTransaction;
